File: ts/project/.sdk/tm/zig/core/context.zig
// Operation context (mirrors go core/context.go / rust core/context.rs).
// Contexts are heap-allocated (*Context) and mutated in place through the
// pointer — Zig needs no RefCell, and the Value fields are already
// reference-stable via *MapRef / *ListRef.

const std = @import("std");
const h = @import("helpers.zig");
const err = @import("error.zig");
const control_mod = @import("control.zig");
const operation_mod = @import("operation.zig");
const spec_mod = @import("spec.zig");
const response_mod = @import("response.zig");
const result_mod = @import("result.zig");
const types = @import("types.zig");
const utility_mod = @import("utility.zig");
const sdk = @import("sdk.zig");

const Value = h.Value;
const Control = control_mod.Control;
const Operation = operation_mod.Operation;
const Spec = spec_mod.Spec;
const Response = response_mod.Response;
const SdkResult = result_mod.SdkResult;
const Entity = types.Entity;
const OutVal = types.OutVal;
const Utility = utility_mod.Utility;
const SdkError = err.ProjectNameError;

pub const OpMap = std.StringHashMap(*Operation);

// Construction spec for a Context.
pub const CtxSpec = struct {
    opname: ?[]const u8 = null,
    client: ?*sdk.ProjectNameSDK = null,
    utility: ?*Utility = null,
    ctrl: ?Value = null,
    ctrl_obj: ?*Control = null,
    meta: ?Value = null,
    config: ?Value = null,
    entopts: ?Value = null,
    options: ?Value = null,
    entity: ?Entity = null,
    shared: ?Value = null,
    opmap: ?*OpMap = null,
    data: ?Value = null,
    reqdata: ?Value = null,
    mtch: ?Value = null,
    reqmatch: ?Value = null,
    point: ?Value = null,
    spec: ?*Spec = null,
    result: ?*SdkResult = null,
    response: ?*Response = null,
};

pub const Context = struct {
    id: []const u8,
    out: std.StringHashMap(OutVal),
    ctrl: *Control,
    meta: Value,
    client: ?*sdk.ProjectNameSDK,
    utility: ?*Utility,
    op: *Operation,
    point: Value,
    config: Value,
    entopts: Value,
    options: Value,
    opmap: *OpMap,
    response: ?*Response,
    result: ?*SdkResult,
    spec: ?*Spec,
    data: Value,
    reqdata: Value,
    mtch: Value,
    reqmatch: Value,
    entity: ?Entity,
    shared: Value,
    pending_err: ?*SdkError = null,

    fn mapOrEmpty(v: ?Value) Value {
        if (v) |val| {
            if (h.to_map(val) == .object) return val;
        }
        return h.omap();
    }

    pub fn new(ctxspec: CtxSpec, basectx: ?*Context) *Context {
        const id = std.fmt.allocPrint(h.A(), "C{d}", .{h.rand_int(90000000) + 10000000}) catch "C0";

        const client = ctxspec.client orelse (if (basectx) |b| b.client else null);
        const utility = ctxspec.utility orelse (if (basectx) |b| b.utility else null);

        // Ctrl
        var ctrl: *Control = undefined;
        if (ctxspec.ctrl) |cm| {
            const c = Control.make();
            if (h.get_bool(cm, "throw")) |t| c.throw = t;
            if (h.getp(cm, "explain") == .object) c.explain = h.getp(cm, "explain");
            if (h.get_str(cm, "actor")) |a| c.actor = a;
            if (h.getp(cm, "paging") == .object) c.paging = h.getp(cm, "paging");
            ctrl = c;
        } else if (ctxspec.ctrl_obj) |co| {
            ctrl = co;
        } else if (basectx) |b| {
            ctrl = b.ctrl;
        } else {
            ctrl = Control.make();
        }

        // Meta
        const meta: Value = ctxspec.meta orelse blk: {
            if (basectx) |b| {
                if (b.meta == .object) break :blk b.meta;
            }
            break :blk h.omap();
        };

        // Config / Entopts / Options / Entity / Shared: fall back to base.
        const config: Value = ctxspec.config orelse (if (basectx) |b| b.config else h.vnull());
        const entopts: Value = ctxspec.entopts orelse (if (basectx) |b| b.entopts else h.vnull());
        const options: Value = ctxspec.options orelse (if (basectx) |b| b.options else h.vnull());
        const entity: ?Entity = ctxspec.entity orelse (if (basectx) |b| b.entity else null);
        const shared: Value = ctxspec.shared orelse blk: {
            if (basectx) |b| {
                if (b.shared == .object) break :blk b.shared;
            }
            break :blk h.vnull();
        };

        // Opmap (shared with the base context).
        const opmap: *OpMap = ctxspec.opmap orelse (if (basectx) |b| b.opmap else blk: {
            const m = h.A().create(OpMap) catch unreachable;
            m.* = OpMap.init(h.A());
            break :blk m;
        });

        // Data maps (never inherited).
        const data = mapOrEmpty(ctxspec.data);
        const reqdata = mapOrEmpty(ctxspec.reqdata);
        const mtch = mapOrEmpty(ctxspec.mtch);
        const reqmatch = mapOrEmpty(ctxspec.reqmatch);

        // Point / Spec / Result / Response: fall back to base.
        const point: Value = ctxspec.point orelse blk: {
            if (basectx) |b| {
                if (b.point == .object) break :blk b.point;
            }
            break :blk h.vnull();
        };
        const spec: ?*Spec = ctxspec.spec orelse (if (basectx) |b| b.spec else null);
        const result: ?*SdkResult = ctxspec.result orelse (if (basectx) |b| b.result else null);
        const response: ?*Response = ctxspec.response orelse (if (basectx) |b| b.response else null);

        const ctx = h.A().create(Context) catch unreachable;
        ctx.* = .{
            .id = id,
            .out = std.StringHashMap(OutVal).init(h.A()),
            .ctrl = ctrl,
            .meta = meta,
            .client = client,
            .utility = utility,
            .op = Operation.make(h.omap()),
            .point = point,
            .config = config,
            .entopts = entopts,
            .options = options,
            .opmap = opmap,
            .response = response,
            .result = result,
            .spec = spec,
            .data = data,
            .reqdata = reqdata,
            .mtch = mtch,
            .reqmatch = reqmatch,
            .entity = entity,
            .shared = shared,
            .pending_err = null,
        };

        const opname = ctxspec.opname orelse "";
        ctx.op = ctx.resolve_op(opname);

        return ctx;
    }

    fn resolve_op(self: *Context, opname: []const u8) *Operation {
        const entname: []const u8 = if (self.entity) |e| e.get_name() else "";
        const cache_key = std.fmt.allocPrint(h.A(), "{s}:{s}", .{ entname, opname }) catch opname;

        if (self.opmap.get(cache_key)) |op| return op;

        if (opname.len == 0) {
            return Operation.make(h.omap());
        }

        const opcfg = h.getpath(&.{ "entity", entname, "op", opname }, self.config);

        const input: []const u8 = if (std.mem.eql(u8, opname, "update") or std.mem.eql(u8, opname, "create")) "data" else "match";

        const targets: Value = switch (h.getp(opcfg, "points")) {
            .array => h.getp(opcfg, "points"),
            else => h.olist(),
        };

        const op = Operation.make(h.jo(&.{
            .{ "entity", h.vstr(entname) },
            .{ "name", h.vstr(opname) },
            .{ "input", h.vstr(input) },
            .{ "points", targets },
        }));

        self.opmap.put(cache_key, op) catch {};
        return op;
    }

    pub fn make_error(self: *Context, code: []const u8, msg: []const u8) *SdkError {
        _ = self;
        return SdkError.make(code, msg);
    }

    // Stash a pending error and return the pipeline error sentinel.
    pub fn fail(self: *Context, code: []const u8, msg: []const u8) err.E {
        self.pending_err = SdkError.make(code, msg);
        return error.Sdk;
    }

    pub fn fail_err(self: *Context, e: *SdkError) err.E {
        self.pending_err = e;
        return error.Sdk;
    }

    pub fn take_err(self: *Context) ?*SdkError {
        const e = self.pending_err;
        self.pending_err = null;
        return e;
    }

    pub fn util(self: *Context) *Utility {
        return self.utility orelse unreachable;
    }

    // --- ctx.out staging helpers ---

    pub fn out_get(self: *Context, key: []const u8) ?OutVal {
        return self.out.get(key);
    }
    pub fn out_set(self: *Context, key: []const u8, val: OutVal) void {
        self.out.put(key, val) catch {};
    }
    pub fn out_val(self: *Context, key: []const u8) Value {
        if (self.out.get(key)) |ov| {
            switch (ov) {
                .val => |v| return v,
                else => {},
            }
        }
        return h.vnull();
    }
};

File: ts/project/.sdk/tm/c/core/context.c
// Operation context (mirrors core/context.rs). Plain heap struct; single
// threaded, never-free — fields mutated in place.

#include "sdk.h"

#include <stdlib.h>
#include <string.h>
#include <stdio.h>

static char* dup_str(const char* s) {
  if (!s) s = "";
  size_t n = strlen(s);
  char* d = (char*)malloc(n + 1);
  memcpy(d, s, n + 1);
  return d;
}

PNError* context_make_error(Context* ctx, const char* code, const char* msg) {
  (void)ctx;
  return pn_error_new(code, msg);
}

Utility* context_util(Context* ctx) { return ctx->utility; }

void ctx_out_set_point_val(Context* ctx, voxgig_value* v) {
  ctx->out_point_kind = OUT_VAL;
  ctx->out_point_val = v;
  ctx->out_point_err = NULL;
}
void ctx_out_set_point_err(Context* ctx, PNError* e) {
  ctx->out_point_kind = OUT_ERR;
  ctx->out_point_err = e;
  ctx->out_point_val = NULL;
}
voxgig_value* ctx_out_extra_get(Context* ctx, const char* key) {
  return getp(ctx->out_extra, key);
}
void ctx_out_extra_set(Context* ctx, const char* key, voxgig_value* v) {
  setp(ctx->out_extra, key, v);
}

static Operation* resolve_op(Context* ctx, const char* opname) {
  const char* entname = ctx->entity ? ctx->entity->vt->get_name(ctx->entity) : "";

  if (opname == NULL || opname[0] == '\0') {
    return operation_new(voxgig_new_map());
  }

  voxgig_value* opcfg;
  {
    const char* keys[5] = {"entity", entname, "op", opname, NULL};
    opcfg = getpath_c(ctx->config, keys);
  }

  const char* input =
      (strcmp(opname, "update") == 0 || strcmp(opname, "create") == 0) ? "data" : "match";

  voxgig_value* targets = getp(opcfg, "points");
  if (!voxgig_is_list(targets)) targets = voxgig_new_list();

  voxgig_value* opmap = cmap(4, "entity", v_str(entname), "name", v_str(opname),
                            "input", v_str(input), "points", targets);
  return operation_new(opmap);
}

Context* context_new(CtxSpec cs, Context* basectx) {
  Context* ctx = (Context*)calloc(1, sizeof(Context));

  char idbuf[32];
  snprintf(idbuf, sizeof(idbuf), "C%lld", (long long)(rand_int(90000000) + 10000000));
  ctx->id = dup_str(idbuf);

  ctx->out_point_kind = OUT_NONE;
  ctx->out_extra = voxgig_new_map();

  // Client.
  ctx->client = cs.client ? cs.client : (basectx ? basectx->client : NULL);
  // Utility.
  ctx->utility = cs.utility ? cs.utility : (basectx ? basectx->utility : NULL);

  // Ctrl.
  if (cs.ctrl) {
    Control* c = control_new();
    bool t;
    if (get_bool(cs.ctrl, "throw", &t)) { c->has_throw = true; c->throw_v = t; }
    voxgig_value* explain = getp(cs.ctrl, "explain");
    if (voxgig_is_map(explain)) c->explain = explain;
    const char* actor = get_str(cs.ctrl, "actor");
    if (actor) { free(c->actor); c->actor = dup_str(actor); }
    voxgig_value* paging = getp(cs.ctrl, "paging");
    if (voxgig_is_map(paging)) c->paging = paging;
    ctx->ctrl = c;
  } else if (cs.ctrl_obj) {
    ctx->ctrl = cs.ctrl_obj;
  } else if (basectx) {
    ctx->ctrl = basectx->ctrl;
  } else {
    ctx->ctrl = control_new();
  }

  // Meta.
  if (cs.meta) {
    ctx->meta = cs.meta;
  } else if (basectx && voxgig_is_map(basectx->meta)) {
    ctx->meta = basectx->meta;
  } else {
    ctx->meta = voxgig_new_map();
  }

  // Config / entopts / options / entity / shared: fall back to base.
  ctx->config = cs.config ? cs.config : (basectx ? basectx->config : voxgig_new_undef());
  ctx->entopts = cs.entopts ? cs.entopts : (basectx ? basectx->entopts : voxgig_new_undef());
  ctx->options = cs.options ? cs.options : (basectx ? basectx->options : voxgig_new_undef());
  ctx->entity = cs.entity ? cs.entity : (basectx ? basectx->entity : NULL);
  if (cs.shared) {
    ctx->shared = cs.shared;
  } else if (basectx && voxgig_is_map(basectx->shared)) {
    ctx->shared = basectx->shared;
  } else {
    ctx->shared = voxgig_new_undef();
  }

  // Data maps (never inherited).
  ctx->data = (cs.data && voxgig_is_map(cs.data)) ? cs.data : voxgig_new_map();
  ctx->reqdata = (cs.reqdata && voxgig_is_map(cs.reqdata)) ? cs.reqdata : voxgig_new_map();
  ctx->mtch = (cs.mtch && voxgig_is_map(cs.mtch)) ? cs.mtch : voxgig_new_map();
  ctx->reqmatch = (cs.reqmatch && voxgig_is_map(cs.reqmatch)) ? cs.reqmatch : voxgig_new_map();

  // Point / spec / result / response: fall back to base.
  if (cs.point) {
    ctx->point = cs.point;
  } else if (basectx && voxgig_is_map(basectx->point)) {
    ctx->point = basectx->point;
  } else {
    ctx->point = voxgig_new_undef();
  }
  ctx->spec = cs.spec ? cs.spec : (basectx ? basectx->spec : NULL);
  ctx->result = cs.result ? cs.result : (basectx ? basectx->result : NULL);
  ctx->response = cs.response ? cs.response : (basectx ? basectx->response : NULL);

  // Resolve operation.
  ctx->op = resolve_op(ctx, cs.opname);

  return ctx;
}
